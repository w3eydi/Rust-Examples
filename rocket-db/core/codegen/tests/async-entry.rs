#![allow(dead_code, unused_variables)]

mod a {
    // async launch that is async.
    #[rocket::launch]
    async fn rocket() -> rocket::Rocket<rocket::Build> {
        let _ = rocket::build().launch().await;
        rocket::build()
    }

    async fn use_it() {
        let rocket: rocket::Rocket<rocket::Build> = rocket().await;
    }
}

mod b {
    // async launch that isn't async.
    #[rocket::launch]
    async fn main2() -> _ {
        rocket::build()
    }

    async fn use_it() {
        let rocket: rocket::Rocket<_> = main2().await;
    }
}

mod b_inferred {
    #[rocket::launch]
    async fn main2() -> _ { rocket::build() }

    async fn use_it() {
        let rocket: rocket::Rocket<_> = main2().await;
    }
}

mod c {
    // non-async launch.
    #[rocket::launch]
    fn rocket() -> _ {
        rocket::build()
    }

    fn use_it() {
        let rocket: rocket::Rocket<_> = rocket();
    }
}

mod c_inferred {
    #[rocket::launch]
    fn rocket() -> _ { rocket::build() }

    fn use_it() {
        let rocket: rocket::Rocket<_> = rocket();
    }
}

mod d {
    // main with async, is async.
    #[rocket::main]
    async fn main() {
        let _ = rocket::build().launch().await;
    }
}

mod e {
    // main with async, isn't async.
    #[rocket::main]
    async fn main() { }
}

mod f {
    // main with async, is async, with termination return.
    #[rocket::main]
    async fn main() -> Result<(), String> {
        let result = rocket::build().launch().await;
        result.map_err(|e| e.to_string())
    }
}

mod g {
    // main with async, isn't async, with termination return.
    #[rocket::main]
    async fn main() -> Result<(), String> {
        Ok(())
    }
}

// main with async, is async, with termination return.
#[rocket::main]
async fn main() -> Result<(), String> {
    let result = rocket::build().launch().await;
    result.map_err(|e| e.to_string())
}
