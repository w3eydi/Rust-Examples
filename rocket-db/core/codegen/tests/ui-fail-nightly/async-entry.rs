../ui-fail/async-entry.rs