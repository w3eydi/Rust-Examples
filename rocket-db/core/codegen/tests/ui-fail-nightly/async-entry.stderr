error: attribute can only be applied to `async` functions
 --> $DIR/async-entry.rs:4:5
  |
4 |     #[rocket::main]
  |     ^^^^^^^^^^^^^^^
  |
note: this function must be `async`
 --> $DIR/async-entry.rs:5:5
  |
5 |     fn foo() { }
  |     ^^^^^^^^
  = note: this error originates in the attribute macro `rocket::main` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: attribute is typically applied to `main` function
  --> $DIR/async-entry.rs:10:5
   |
10 |     #[rocket::main]
   |     ^^^^^^^^^^^^^^^
   |
note: this function is not `main`
  --> $DIR/async-entry.rs:11:14
   |
11 |     async fn foo() { }
   |              ^^^
   = note: this warning originates in the attribute macro `rocket::main` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute can only be applied to `async` functions
  --> $DIR/async-entry.rs:16:5
   |
16 |     #[rocket::main]
   |     ^^^^^^^^^^^^^^^
   |
note: this function must be `async`
  --> $DIR/async-entry.rs:17:5
   |
17 |     fn main() {
   |     ^^^^^^^^^
   = note: this error originates in the attribute macro `rocket::main` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute cannot be applied to `main` function
  --> $DIR/async-entry.rs:49:5
   |
49 |     #[rocket::launch]
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: this attribute generates a `main` function
note: this function cannot be `main`
  --> $DIR/async-entry.rs:50:8
   |
50 |     fn main() -> rocekt::Rocket<rocket::Build> {
   |        ^^^^
   = note: this error originates in the attribute macro `rocket::launch` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute can only be applied to functions that return a value
  --> $DIR/async-entry.rs:56:5
   |
56 |     #[rocket::launch]
   |     ^^^^^^^^^^^^^^^^^
   |
note: this function must return a value
  --> $DIR/async-entry.rs:57:5
   |
57 |     async fn rocket() {
   |     ^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `rocket::launch` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute can only be applied to functions that return a value
  --> $DIR/async-entry.rs:64:5
   |
64 |     #[rocket::launch]
   |     ^^^^^^^^^^^^^^^^^
   |
note: this function must return a value
  --> $DIR/async-entry.rs:65:5
   |
65 |     fn rocket() {
   |     ^^^^^^^^^^^
   = note: this error originates in the attribute macro `rocket::launch` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute cannot be applied to `main` function
  --> $DIR/async-entry.rs:79:5
   |
79 |     #[rocket::launch]
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: this attribute generates a `main` function
note: this function cannot be `main`
  --> $DIR/async-entry.rs:80:8
   |
80 |     fn main() -> &'static str {
   |        ^^^^
   = note: this error originates in the attribute macro `rocket::launch` (in Nightly builds, run with -Z macro-backtrace for more info)

error: attribute cannot be applied to `main` function
  --> $DIR/async-entry.rs:87:5
   |
87 |     #[rocket::launch]
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: this attribute generates a `main` function
note: this function cannot be `main`
  --> $DIR/async-entry.rs:88:14
   |
88 |     async fn main() -> _ {
   |              ^^^^
   = note: this error originates in the attribute macro `rocket::launch` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0728]: `await` is only allowed inside `async` functions and blocks
  --> $DIR/async-entry.rs:73:17
   |
72 |     fn rocket() -> _ {
   |        ------ this is not `async`
73 |         let _ = rocket::build().launch().await;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ only allowed inside `async` functions and blocks

error[E0308]: mismatched types
  --> $DIR/async-entry.rs:35:9
   |
35 |         rocket::build()
   |         ^^^^^^^^^^^^^^^ expected struct `std::string::String`, found struct `Rocket`
   |
   = note: expected struct `std::string::String`
              found struct `Rocket<Build>`

error[E0308]: mismatched types
  --> $DIR/async-entry.rs:44:9
   |
44 |         "hi".to_string()
   |         ^^^^^^^^^^^^^^^^ expected struct `Rocket`, found struct `std::string::String`
   |
   = note: expected struct `Rocket<Build>`
              found struct `std::string::String`

error[E0308]: mismatched types
  --> $DIR/async-entry.rs:24:21
   |
24 |       async fn main() {
   |                       ^ expected `()` because of default return type
   |  _____________________|
   | |
25 | |         rocket::build()
26 | |     }
   | |     ^- help: consider using a semicolon here: `;`
   | |_____|
   |       expected `()`, found struct `Rocket`
   |
   = note: expected unit type `()`
                 found struct `Rocket<Build>`

error[E0308]: mismatched types
  --> $DIR/async-entry.rs:33:26
   |
33 |     async fn rocket() -> String {
   |                          ^^^^^^
   |                          |
   |                          expected struct `Rocket`, found struct `std::string::String`
   |                          expected due to this
   |
   = note: expected struct `Rocket<Build>`
              found struct `std::string::String`

error[E0277]: `main` has invalid return type `Rocket<Build>`
  --> $DIR/async-entry.rs:94:20
   |
94 | async fn main() -> rocket::Rocket<rocket::Build> {
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `main` can only return types that implement `Termination`
   |
   = help: consider using `()`, or a `Result`
